cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/YarnLoadBalancerLaunchService.java

```diff
@@ -0,0 +1,264 @@
+        return gatewayIPs;
+    }
+
+    /**
+     * Gets all of the containers currently running in the YCloud ecosystem within the application
+     * given by the application name provided.
+     */
+    public Iterable<Container> getContainers(String applicationName, YarnClient yarnClient) {
+        LOGGER.debug("Getting the Yarn containers for application " + applicationName + ".");
+        ApplicationDetailRequest applicationDetailRequest = new ApplicationDetailRequest();
+        applicationDetailRequest.setName(applicationName);
+        ResponseContext responseContext;
+
+        try {
+            responseContext = yarnClient.getApplicationDetail(applicationDetailRequest);
+            LOGGER.debug("Successfully for a response for Yarn application " + applicationName + " from the Yarn client.");
+        } catch (MalformedURLException ex) {
+            LOGGER.error("Failed to get information for the Yarn loadbalancer! Application name: " + applicationName + " Error: " + ex.getMessage());
+            throw new CloudConnectorException("Failed to get information for the Yarn loadbalancer.", ex);
+        }
+
+        if (responseContext.getStatusCode() == YarnResourceConstants.HTTP_SUCCESS) {
+            LOGGER.debug("Successfully retrieved container information for the Yarn application " + applicationName + ".");
+            ApplicationDetailResponse applicationDetailResponse = (ApplicationDetailResponse) responseContext.getResponseObject();
+            return applicationDetailResponse.getContainers();
+        } else {
+            LOGGER.error("Failed to get yarn container! Application name: " + applicationName);
+            ApplicationErrorResponse errorResponse = responseContext.getResponseError();
+            throw new CloudConnectorException(String.format("Failed to get yarn container details: HTTP Return: %d Error: %s",
+                    responseContext.getStatusCode(), errorResponse == null ? "unknown" : errorResponse.getDiagnostics()));
+        }
+    }
+
+    /**
+     * Gets the names of the instance groups that are Gateway types, the names can then be compared against
+     * Yarn container names to check whether a container is a gateway container or not.
+     */
+    private Set<String> getGatewayGroupNames(CloudStack cloudStack) {
+        return cloudStack.getGroups().stream().filter(group -> InstanceGroupType.isGateway(group.getType()))
+                .map(Group::getName).collect(Collectors.toSet());
+    }
+
+    /**
+     * Creates a Yarn component for each of the loadbalancers that are to be created, using a set of pre-defined
+     * constants for the various component parameters.
+     *
+     * Each container is pointed at a specific docker image which contains the loadbalancing service and logic.
+     */
+    private List<YarnComponent> createLoadBalancerComponents(CloudStack cloudStack, List<String> gatewayIPs, String applicationName) {
```

Comment:
I would move the creation of the components into an individual service/factory. The `YarnLoadBalancerLaunchService.java` would be much more clean.